src/plugins/intel_gpu/src/runtime/kernels_cache.cpp

```diff
@@ -90,48 +89,58 @@ void kernels_cache::get_program_source(const kernels_code& kernels_source_code,
+            std::string entry_point = kernel_string->entry_point;
+            std::string options = kernel_string->options;
+            bool batch_compilation = kernel_string->batch_compilation;
 
-        std::string key = options;
+            if (batch_compilation) {
+                options = reorder_options(options);
+            }
 
-        if (batch_compilation == false) {
-            key += " __PROGRAM__" + std::to_string(program_buckets.size());
-        }
+            std::string key = options;
 
-        if (dump_custom_program) {
-            key += " __DUMP_CUSTOM_PROGRAM__";  // Adding label to key so it would be separated from other programs
-        }
+            if (batch_compilation == false) {
+                key += " __PROGRAM__" + std::to_string(program_buckets.size());
+            }
 
-        auto& bucket_id = std::get<0>(program_buckets[key]);
-        auto& current_bucket = std::get<1>(program_buckets[key]);
-        if (current_bucket.empty()) { // new bucket
-            const auto& batch_id = 0;
-            // increase bucket id if and only if new bucket comes
-            bucket_id = static_cast<int32_t>(program_buckets.size() - 1);
-            current_bucket.push_back(batch_program(bucket_id, batch_id, options, batch_header_str));
-        }
+            if (dump_custom_program) {
+                key += " __DUMP_CUSTOM_PROGRAM__";  // Adding label to key so it would be separated from other programs
+            }
 
-        // Create new kernels batch when the limit is reached
-        if (current_bucket.back().kernels_counter >= get_max_kernels_per_batch()) {
-            const auto& batch_id = static_cast<int32_t>(current_bucket.size());
-            current_bucket.push_back(batch_program(bucket_id, batch_id, options, batch_header_str));
-        }
+            auto& bucket_id = std::get<0>(program_buckets[key]);
+            auto& current_bucket = std::get<1>(program_buckets[key]);
+            if (current_bucket.empty()) { // new bucket
+                const auto& batch_id = 0;
+                // increase bucket id if and only if new bucket comes
+                bucket_id = static_cast<int32_t>(program_buckets.size() - 1);
+                current_bucket.push_back(batch_program(bucket_id, batch_id, options, batch_header_str));
+            }
+
+            // Create new kernels batch when the limit is reached
+            // and current kernel's entry_point is duplicated in this kernels batch
+            if (current_bucket.back().kernels_counter >= get_max_kernels_per_batch()
+                || std::find(entry_points.begin(), entry_points.end(), entry_point) != entry_points.end()) {
+                const auto& batch_id = static_cast<int32_t>(current_bucket.size());
+                current_bucket.push_back(batch_program(bucket_id, batch_id, options, batch_header_str));
+                entry_points.clear();
```

Comment:
Where is the duplication of kernels handled? e.g., two kernels with same impl_param, same entry. 
Should't we distribute batches like 
1. scan the entry point and check dupilication 
- [entry_point_A] : {kernel_from impl_param_A_1), kernel_from_impl_param_A_2 }
- [entry_point_B] : kernel_from_impl_param_B_1
- ...  
2. Distribute kernel_from_impl_param_A_1 and kernel_from_impl_param_A_2 to different batches 
- min # of batches : max # of bucket size of above.

Reply:
I think duplication checking code is missed during code rebase. I will update it.